src/jupyterlabcontroller/services/builder/lab.py

```diff
@@ -0,0 +1,821 @@
+            ),
+        )
+
+    def _build_pod_init_containers(
+        self, user: GafaelfawrUserInfo, resources: LabResources
+    ) -> list[V1Container]:
+        """Build init containers for the pod."""
+        username = user.username
+        as_root = V1SecurityContext(
+            allow_privilege_escalation=True,
+            run_as_non_root=False,
+            run_as_user=0,
+        )
+        as_user = V1SecurityContext(
+            allow_privilege_escalation=False,
+            run_as_non_root=True,
+            run_as_user=user.uid,
+        )
+
+        # Use the same environment ConfigMap as the notebook container since
+        # it may contain other things we need for provisioning. Add
+        # environment variables communicating the primary UID and GID, which
+        # is our main interface to provisioning init containers.
+        env_source = V1ConfigMapEnvSource(name=f"{username}-nb-env")
+        env = [
+            V1EnvVar(name="EXTERNAL_GID", value=str(user.gid)),
+            V1EnvVar(name="EXTERNAL_UID", value=str(user.uid)),
+        ]
+
+        containers = []
+        for spec in self._config.init_containers:
+            volumes = self.build_lab_config_volumes(username, spec.volumes)
+            container = V1Container(
+                name=spec.name,
+                env=env,
+                env_from=[V1EnvFromSource(config_map_ref=env_source)],
+                image=spec.image,
+                resources=resources.to_kubernetes(),
+                security_context=as_root if spec.privileged else as_user,
+                volume_mounts=[v.volume_mount for v in volumes],
+            )
+            containers.append(container)
+
+        return containers
+
+    def _build_pod_containers(
+        self,
+        user: GafaelfawrUserInfo,
+        mounts: list[V1VolumeMount],
+        resources: LabResources,
+        image: RSPImage,
+    ) -> V1PodSpec:
+        """Construct the containers for the user's lab pod."""
+        # Additional environment variables to set, layered on top of the env
+        # ConfigMap. The ConfigMap holds public information known before the
+        # lab is spawned; these environment variables hold everything else.
+        env = [
+            # User's Gafaelfawr token.
+            V1EnvVar(
+                name="ACCESS_TOKEN",
+                value_from=V1EnvVarSource(
+                    secret_key_ref=V1SecretKeySelector(
+                        key="token", name=f"{user.username}-nb", optional=False
+                    )
+                ),
+            ),
+            # Node on which the pod is running.
+            V1EnvVar(
+                name="KUBERNETES_NODE_NAME",
+                value_from=V1EnvVarSource(
+                    field_ref=V1ObjectFieldSelector(field_path="spec.nodeName")
+                ),
+            ),
+            # Deprecated version of KUBERNETES_NODE_NAME, used by lsst.rsp
+            # 0.3.4 and earlier.
```

Comment:
We really need to prioritize a reaper for old images, so that we know when anything that uses 0.3.4 has gone away, so we can clean up this (and other) tech debt items.